core/builder/src/main/java/io/quarkus/builder/Execution.java

```diff
@@ -39,8 +45,14 @@
     private volatile Thread runningThread;
     private volatile boolean done;
 
+    /**
+     * A topological ordering of {@link StepInfo}s.
+     */
+    private final Map<StepInfo, Integer> stepOrdinals;
```

Comment:
Rather than have maps associating things with `StepInfo`, why not just put fields right on the `StepInfo`?

Reply:
Good point, thanks. I originally thought I'd have to break `StepInfo`'s immutability to do that, but now checking again it looks like it is doable.

Reply:
`StepInfo` now has an `ordinal` field in aacdcf7.